test/integration/tracked_vehicle_system.cc

```diff
@@ -0,0 +1,355 @@
+#include "ignition/gazebo/Server.hh"
+#include "ignition/gazebo/test_config.hh"
+
+#include "../helpers/Relay.hh"
+
+#define tol 10e-4
+
+using namespace ignition;
+using namespace gazebo;
+using namespace std::chrono_literals;
+
+// Verify that a model's world pose is near a specified pose.
+void verifyPose(const math::Pose3d& pose1, const math::Pose3d& pose2)
+{
+  EXPECT_NEAR(pose1.Pos().X(), pose2.Pos().X(), 1e-1);
+  EXPECT_NEAR(pose1.Pos().Y(), pose2.Pos().Y(), 1e-1);
+  EXPECT_NEAR(pose1.Pos().Z(), pose2.Pos().Z(), 1e-2);
+  EXPECT_NEAR(pose1.Rot().Roll(), pose2.Rot().Roll(), 1e-2);
+  EXPECT_NEAR(pose1.Rot().Pitch(), pose2.Rot().Pitch(), 1e-2);
+  EXPECT_NEAR(pose1.Rot().Yaw(), pose2.Rot().Yaw(), 1e-1);
+}
+
+/// \brief Test TrackedVehicle system. This test drives a tracked robot over a
+/// course of obstacles and verifies that it is able to climb on/over them.
+class TrackedVehicleTest : public ::testing::TestWithParam<int>
+{
+  // Documentation inherited
+  protected: void SetUp() override
+  {
+    common::Console::SetVerbosity(4);
+    ignition::common::setenv("IGN_GAZEBO_SYSTEM_PLUGIN_PATH",
+           (std::string(PROJECT_BINARY_PATH) + "/lib").c_str());
+  }
+
+  /// \param[in] _sdfFile SDF file to load.
+  /// \param[in] _cmdVelTopic Command velocity topic.
+  /// \param[in] _odomTopic Odometry topic.
+  protected: void TestPublishCmd(const std::string &_sdfFile,
+                                 const std::string &_cmdVelTopic,
+                                 const std::string &_odomTopic)
+  {
+    // Start server
+    ServerConfig serverConfig;
+    serverConfig.SetSdfFile(_sdfFile);
+
+    Server server(serverConfig);
+    EXPECT_FALSE(server.Running());
+    EXPECT_FALSE(*server.Running(0));
+
+    // Create a system that records the vehicle poses
+    test::Relay testSystem;
+    test::Relay ecmGetterSystem;
+
+    Entity modelEntity {kNullEntity};
+    std::vector<math::Pose3d> poses;
+    testSystem.OnPostUpdate([&](const gazebo::UpdateInfo &,
+      const gazebo::EntityComponentManager &_ecm)
+      {
+        modelEntity = _ecm.EntityByComponents(
+          components::Model(),
+          components::Name("simple_tracked"));
+        EXPECT_NE(kNullEntity, modelEntity);
+
+        auto poseComp = _ecm.Component<components::Pose>(modelEntity);
+        ASSERT_NE(nullptr, poseComp);
+
+        poses.push_back(poseComp->Data());
+      });
+    server.AddSystem(testSystem.systemPtr);
+
+    EntityComponentManager* ecm {nullptr};
+    ecmGetterSystem.OnPreUpdate([&ecm](const gazebo::UpdateInfo &,
+      gazebo::EntityComponentManager &_ecm)
+      {
+        if (ecm == nullptr)
+          ecm = &_ecm;
+      });
+    server.AddSystem(ecmGetterSystem.systemPtr);
+
+    // Run server and check that vehicle didn't move
+    server.Run(true, 1000, false);
+
+    EXPECT_EQ(1000u, poses.size());
+
+    for (size_t i = 101;  i < poses.size(); ++i)
+    {
+      verifyPose(poses[100], poses[i]);
```

Comment:
Instead of using a function, can you put the function contents (each expectation) here instead? Currently, if one or more fail, it will point to the line numbers in the function (lines 43-48) instead of here

Reply:
Hmm, I wanted to avoid code repetition... I assume a macro would not make the situation better?

Reply:
How about using `SCOPED_TRACE` (https://google.github.io/googletest/advanced.html#adding-traces-to-assertions)?

Reply:
I added the scoped trace and verified it references the right place in source which failed:

```
Google Test trace:
/media/data/ign/ign-gazebo/test/integration/tracked_vehicle_system.cc:115: 993
/media/data/ign/ign-gazebo/test/integration/tracked_vehicle_system.cc:43: Failure
The difference between pose1.Pos().X() and pose2.Pos().X() is 6.0000000000003029, which exceeds 1e-1, where
pose1.Pos().X() evaluates to 3.0000000000062408,
pose2.Pos().X() evaluates to -2.9999999999940621, and
1e-1 evaluates to 0.10000000000000001.

```